### PR TITLE
Error codes string output handling

### LINKED ISSUE #53: error codes are not included in perror output
The error id is not included in the output of `stumpless_perror`, which could add some extra steps when troubleshooting. The goal of the perror function is to prevent manual print statements, and there is also not a way to easily get the name of the error even if the code itself is retrieved (as it will be an integer and not a string).

This could easily be solved by adding a new public function like `stumpless_get_error_id_string` that spits out a string representation of an error id enum. This could then be used in the perror function to add the error id in an easily readable format.

### PR DESCRIPTION
**Pull request by pedestrianlove**
This should fix #53 . 
1. Stringification part
I [used](https://stackoverflow.com/questions/9907160/how-to-convert-enum-names-to-string-in-c) the macro to handle the conversion and it would make sure that the enum and the string are always in sync.
I wrapped the constructed array with a function called `stumpless_get_error_id_string` for a more consistent user experience.

2. Print error part
I put the print error statement between `prefix` and `last_error->message`.

3. Test part
I added the test of string output whenever the `error->message` is being tested.

### CODE REVIEW

**Comment by goatshriek** [comment]
This looks like a solid approach, thank you for putting it together! Here are a few comments to finalize this before it is merged.

Make sure that the doxygen documentation still documents each member of the error_id enum. You can see what this looks like currently [here](https://goatshriek.github.io/stumpless/docs/c/latest/error_8h.html) and can generate it locally using the `docs` target if you have doxygen installed when you run CMake. Using the macro generator approach makes this a little more difficult. You might be able to use the doxygen `/var` command to document the values separate from their actual definition, but this requires you to make changes for each value somewhere else and seems to defeat the purpose of the macro keeping the enum and strings in one place. If you cannot keep the documentation and definitions together, I would reconsider the macro approach as it will add complexity without keeping the definitions to a single place.

You will need to add your new public function `stumpless_get_error_id_string` to the .def file at `src/windows/stumpless.def` to fix the failing windows builds. I apologize - you had no way of knowing this as it is not covered in any current documentation. I have created a new branch `dev-getting-started` with documentation updates to make sure that problems you encounter are covered. [This section](https://github.com/goatshriek/stumpless/blob/dev-getting-started/docs/development.md#adding-new-functions) describes why you are seeing that error. I will add any other gotchas that you encounter to this branch as well. Again I apologize for you having to uncover these shortfalls in documentation.

It appears that there are segmentation faults on certain architectures currently. I'm not sure why you are not seeing them on every architecture, but in any case it is because you have missed a comma in the `GENERATE_STRING` macro which was causing the array to consist of a single long string with every value in it. The output of `strings` on the compiled library reflects this. I created a temporary branch named `pr-99` just to test this on travis, you can check out the latest build [here](https://travis-ci.com/github/goatshriek/stumpless/builds/177437967).

You will see that in `pr-99` I also hardcoded the enumeration values while troubleshooting. This is on our [roadmap](https://github.com/goatshriek/stumpless/blob/latest/docs/roadmap.md#200-next-major-and-minor-release) to increase portability, and this is a good change to incorporate it into if you would like. If not no worries, I will make the change later during 2.0 development.

If you decide to stick with the static array approach, move the definition of the array to `src/error.c` with the other error statics instead of in the public header file. If it is in the header file then the array is created in any program that includes the header as well as the shared library, which is a duplication of values and waste of memory. You can use the `strings` program on an executable or library to see if they are present.

Thanks again for putting in the work on this! It is greatly appreciated.

**Comment by pedestrianlove** [comment]
Ok, I'll try to tackle this one tomorrow.
Thanks for your detailed explanation!

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h]: it turns out that this \var command is not necessary, and can be removed

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h]: rename this to `STUMPLESS_FOREACH_ERROR` so that it does not interfere with the namespace of any files that include this header

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h]: For all multi-line comments within the macro expansion to render properly in doxygen, you will need to remove the asterisks at the beginning of the lines  (for example, lines 39, 40, and 41 here, lines 49-51, etc.)

File: src/error.c
Comment (goatshriek) [inline src/error.c]: This conditional will not actually prevent a segfault if an invalid id is passed in, as it still attempts to use it to index the array. To perform a check on the id you would need to cast it to an int and make sure it is within the valid bounds, something like:
`if( id >= 0 && id < sizeof(stumpless_error_enum_to_string)/sizeof(char *) ){`

File: src/error.c
Comment (goatshriek) [inline src/error.c:57]: You need to add a test to cover the "NO_SUCH_ERROR_ID" scenario, as it is currently [uncovered](https://codecov.io/gh/goatshriek/stumpless/pull/99/diff) and responsible for the pull request check failure you are seeing. I recommend adding the test before you fix the bounds check, so that you see it segfault before the fix and pass afterwards.

File: src/windows/stumpless.def
Comment (goatshriek) [inline src/windows/stumpless.def]: use spaces instead of tabs here to stay consistent with the rest of the file

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h:166]: Document this function (use the formatting of the others in this file as an example) so that it is clear to users how it behaves.

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h]: rename this to `STUMPLESS_GENERATE_ENUM` so that it does not interfere with the namespace of any files that include this header

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h]: either rename this to `STUMPLESS_GENERATE_STRING` for the same reasons as above, or (preferred) move it to `include/private/error.h` as it is not actually needed in the public header.

File: test/function/error.cpp
Comment (goatshriek) [inline test/function/error.cpp:85]: Add your new test to this file somewhere. I would start by copying the `TEST( GetErrorId, Basic )` as a starting point and modifying it to test `stumpless_get_error_id_string` with an invalid id.

File: include/stumpless/error.h
Comment (goatshriek) [inline include/stumpless/error.h:42]: I was able to get documentation generated using your approach, but you will need to update the Doxyfile at `tools/doxygen/Doxyfile.in` in order to make it work. I have included the change below - it consists of moving two settings into the 'deliberate settings' section and making sure that they are both set to `YES`.

```diff
diff --git a/tools/doxygen/Doxyfile.in b/tools/doxygen/Doxyfile.in
index a4b6389..e209077 100644
--- a/tools/doxygen/Doxyfile.in
+++ b/tools/doxygen/Doxyfile.in
@@ -13,6 +13,8 @@ USE_MDFILE_AS_MAINPAGE = @PROJECT_SOURCE_DIR@/README.md
 OPTIMIZE_OUTPUT_FOR_C  = YES
 SEARCH_INCLUDES        = YES
 INCLUDE_PATH           = @PROJECT_SOURCE_DIR@/include @PROJECT_BINARY_DIR@/include
+ENABLE_PREPROCESSING   = YES
+MACRO_EXPANSION        = YES
 
 # default generated settings
 BRIEF_MEMBER_DESC      = YES
@@ -154,8 +156,6 @@ GENERATE_AUTOGEN_DEF   = NO
 GENERATE_PERLMOD       = NO
 PERLMOD_LATEX          = NO
 PERLMOD_PRETTY         = YES
-ENABLE_PREPROCESSING   = YES
-MACRO_EXPANSION        = NO
 EXPAND_ONLY_PREDEF     = NO
 SKIP_FUNCTION_MACROS   = YES
 ALLEXTERNALS           = NO
```

**Review by goatshriek — CHANGES_REQUESTED** [review]
I've added a number of fine-tuning comments to this PR - once these are complete I will merge this. There are quite a few of them, but they are small tweaks for the most part: the addition of the new test to cover the `NO_SUCH_ERROR_ID` case is the biggest update.

**Comment by pedestrianlove** [comment]
@goatshriek Great thanks, I was just working on that macro inline documenting problem at doxygen/doxygen#7887 .
I'll ask another question when I need more info on these requirements.
Great thanks!

File: include/stumpless/error.h
Comment (pedestrianlove) [reply include/stumpless/error.h:166]: Patched now.

**Comment by goatshriek** [comment]
Awesome, sounds good! I will wait to hear from you before I look at this PR again.

File: include/stumpless/error.h
Comment (pedestrianlove) [reply include/stumpless/error.h]: Patched now. I both renamed and moved it io `include/private/error.h`.

**Comment by pedestrianlove** [comment]
@goatshriek 
Hi there, this PR should be good to go if it passed the build. 
I also tried to shrink my garbage commits with git rebase, so that it should be more compact and fit into the [contributing guidelines](https://github.com/goatshriek/stumpless/blob/dev-getting-started/docs/CONTRIBUTING.md#im-helping).

If you have any question or requirement regarding this PR, please feel free to comment again!

Great thanks!

**Comment by pedestrianlove** [comment]
@goatshriek 
Hi there, this PR should be good to go if it passed the build. 
I also tried to shrink my garbage commits with git rebase, so that it should be more compact and fit into the [contributing guidelines](https://github.com/goatshriek/stumpless/blob/dev-getting-started/docs/CONTRIBUTING.md#im-helping).

If you have any question or requirement regarding this PR, please feel free to tag me again!

Great thanks!

**Comment by goatshriek** [comment]
Excellent, thank you once again for putting this together! I have updated the documentation in the project [website](https://goatshriek.github.io/stumpless/) to reflect the changes.

If you have time, please check out the pull request #100 for updates to the documentation based on some of the difficulties you encountered. If you see any other opportunities for improvement or want to make sure I cover something else in that update, please let me know.

**Pull request merged**
